### PR TITLE
Idea: Simplify feature gate checking code

### PR DESCRIPTION
**Pull request by thockin**
This is just an idea that struck me as I was looking at PRs this cycle.  Maybe not worth doing, but was very simple to POC, and IMNSHO is more readable.    These are almost entirely mechanical changes.

Note, this does not fix/change:
* The number of `features` packages which define the individual named gates
* Any place that mutates or passes around the default `FeatureGate` (e.g. `featuregatetesting.SetFeatureGateDuringTest()` is always called with the default gate, but that lives in a generic package; other places inject gates to do testing and I didn't want to make this PR too complex to review).

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
cc @deads2k @logicalhan

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go [hunk 1]
```diff
@@ -31,3 +31,13 @@ var (
...
+
+// Enabled returns true if the key is enabled in the default FeatureGate.
+func Enabled(key featuregate.Feature) bool {
+	return DefaultFeatureGate.Enabled(key)
+}
+
+// KnownFeatures returns a slice of strings describing the default FeatureGate's known features.
+func KnownFeatures() []string {
+	return DefaultFeatureGate.KnownFeatures()
+}
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go:43]: Nice, I like the simplification.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go:43]: It looks like you missed utilfeature->feature rename in `authorizer/rbac/bootstrappolicy/controller_policy.go`

**Comment by thockin** [comment]
I have a couple further cleanups to add just to see what people think, will ping when new commits are pushed

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: That commit didn't convert ALL users, just users that were converted to the package-funcs.  Subsequent commits should get the rest.

**Comment by thockin** [comment]
New push - checking for test fails

**Comment by thockin** [comment]
Fixed lint and build flubs.  Rebased.

**Comment by thockin** [comment]
@dims your fingerprints are on go.mod having a minor version specified - clue me in?

**Comment by dims** [comment]
> @dims your fingerprints are on go.mod having a minor version specified - clue me in?

@thockin explain a bit more please? not sure i understand what you need.

**Comment by thockin** [comment]
@dims

```
$ git blame go.mod | head
f5ea2344ae69 (Jordan Liggitt            2019-04-04 11:51:03 -0400   1) // This is a generated file. Do not edit directly.
fe8ecb18b066 (Qing Ju                   2020-05-31 21:55:48 -0700   2) // Ensure you've carefully read
fe8ecb18b066 (Qing Ju                   2020-05-31 21:55:48 -0700   3) // https://git.k8s.io/community/contributors/devel/sig-architecture/vendor.md
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400   4) // Run hack/pin-dependency.sh to change pinned dependency versions.
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400   5) // Run hack/update-vendor.sh to update go.mod files and the vendor directory.
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400   6) 
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400   7) module k8s.io/kubernetes
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400   8) 
702d911e2260 (Davanum Srinivas          2023-10-18 18:47:37 -0400   9) go 1.21.3
d0261b107702 (Jordan Liggitt            2019-04-01 11:07:04 -0400  10) 
```

Line 9 is you.  

```
$ git show 702d911e2260 | head
commit 702d911e22604aa69b42f77804c27caaab7e4def
gpg: Signature made Wed 25 Oct 2023 01:31:24 PM PDT
gpg:                using DSA key A67E5FD880EB089F2317796780D83A796103BF59
gpg: Can't check signature: No public key
Author: Davanum Srinivas <davanum@gmail.com>
Date:   Wed Oct 18 18:47:37 2023 -0400

    working-config-otel
    
    Signed-off-by: Davanum Srinivas <davanum@gmail.com>
```

But pull-publishing-bot-validate fails:

```
go: errors parsing go.mod:
/home/prow/go/src/k8s.io/kubernetes/go.mod:9: invalid go version '1.21.3': must match format 1.23
```

I admit that I have never seen a 3-segment version number in go.mod, but Go seems to allow it.  What I don't know is a) why?  b) how this is failing for me and not for every PR?

**Comment by thockin** [comment]
Rebased and updated

**Comment by dims** [comment]
@thockin i admit i don't know how that was picked, +1 to drop the minor version. Could be my IDE (`goland`) when i was making changes by hand (and then running our update-vendor.sh) as the otel stuff was trick to figure out from the usual/regular pin-depdendency iterations

**Comment by thockin** [comment]
@dims - I will add a commit here, but if you want to, you can break that out :)

**Comment by thockin** [comment]
hopefully last tests fixed

**Comment by dims** [comment]
> @dims - I will add a commit here, but if you want to, you can break that out :)

+1 to leave it here in this PR :)

**Comment by thockin** [comment]
OK, tests should pass.  I have included a number of "steps" as commits - we can back away from some if people think I took it too far.   This is, overall, a minor cleanup but *I* like it.  I need someone else to positively affirm they think it's worth merging, or else I will abandon it.  Rebasing this one is challenging :)

One open question:
I left "forwarders" in k8s.io/apiserver/pkg/util/feature, which means any new PR could still follow the old pattern.  I'd rather that break, forcing people to adopt the new style, but it's kind of a breaking change. 

We still don't have good ways to write custom linters, do we?

The TL;DR of this PR is this:

What used to read as:
```
if !utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
```

is now:
```
if !featuregate.Enabled(features.MultiCIDRServiceAllocator) {
```

and what used rto be:
```
featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SELinuxMountReadWriteOncePod, false)()
```

is now:
```
defer featuregatetesting.SetFeatureGateDuringTest(t, featuregate.Default, features.SELinuxMountReadWriteOncePod, false)()
```

**Comment by thockin** [comment]
Ha!  Green

**Comment by thockin** [comment]
Rebased and pushed

**Comment by deads2k** [comment]
Having a single instance for the library itself versus an instance for a binary means that some kind of composition become impossible.  In particular, if featuregates are used by libraries (for instance k8s.io/client-go and other-client-type), then composite binaries may become impossible.  Consider a case like

1. k8s.io/client-go has featuregate/Awesome
2. oidc/client has featuregate/Awesome
3. security/some-kind-of-controller depends on k8s.io/client-go and oidc/client
4. at runtime, the featuregate cannot selectively control the kube/Awesome versus oidc/Awesome.

When we can have different instances, it is possible to compose these libraries.  If we have a single instance, we cannot.  To compose the features, we would need to ensure that no one uses the instance that was moved in `Move DefaultFeatureGate to same pkg as gate type`, which suggests we shouldn't move it.

Rather than try to collapse to a single instance, what if we make it easy to compose featuregates from core libraries into the "instance for my binary"?

**Comment by deads2k** [comment]
When writing up how to do client-go featuregates (messy PR here: https://github.com/kubernetes/kubernetes/pull/121122/files#diff-d073b3d85bfc4877d354264d569dd3e19d544d1775a67f1650d661299869aa7aR278), we found a need to have a consumer override default values for gates. This was important even inside of k/k to leverage new client-go features from a controlled set of binaries when no opinion is expressed.

**Comment by logicalhan** [comment]
> Having a single instance for the library itself versus an instance for a binary means that some kind of composition become impossible. In particular, if featuregates are used by libraries (for instance k8s.io/client-go and other-client-type), then composite binaries may become impossible. Consider a case like
> 
> 1. k8s.io/client-go has featuregate/Awesome
> 2. oidc/client has featuregate/Awesome
> 3. security/some-kind-of-controller depends on k8s.io/client-go and oidc/client
> 4. at runtime, the featuregate cannot selectively control the kube/Awesome versus oidc/Awesome.
> 
> When we can have different instances, it is possible to compose these libraries. If we have a single instance, we cannot. To compose the features, we would need to ensure that no one uses the instance that was moved in `Move DefaultFeatureGate to same pkg as gate type`, which suggests we shouldn't move it.
> 
> Rather than try to collapse to a single instance, what if we make it easy to compose featuregates from core libraries into the "instance for my binary"?

I'm not sure I follow... `NewFeatureGate()` is still exposed, so you can still create new gates and allow for composability. `DefaultMutableFeatureGate` and `DefaultFeatureGate` are also already globals and all @thockin is doing here is making global convenience functions around these already global vars. Any non-default gates would still have `Enabled()`, `Add()` and the rest of the flag methods.

**Comment by thockin** [comment]
Mostly I am trying to streamline what is already being done.

On one hand, if you compose binaries, don't you WANT the same flags for all of them?  Is there a real reason to have them differ?  Isn't this what the gates-as-API proposal is trying to do? 

On the other hand, if every binary wants to pass around a FeatureGate object *everywhere*, they can.  We have this global `DefaultFeatureGate` ostensibly because it is too onerous and not worth the cost to do literally per-binary.  The same argument is often made for logging libs.  If we want to encourage a per-org "global default" (not really per-binary, since we have a spiderweb of deps that need to find it), we probably shouldn't call it "Default" and it probably should not live under "apiserver"?  

If you want to drop this last part and then revisit, we can, but EVERYTHING links "k8s.io/apiserver/pkg/util/feature" and references `DefaultFeatureGate` today, so what you are calling for is a pretty huge undertaking.

**Comment by thockin** [comment]
@logicalhan I think @deads2k point is that putting the global `DefaultFeatureGate` in "k8s.io/component-base/featuregate" instead of "k8s.io/apiserver/pkg/util/feature" makes it a bit more "official".

**Comment by logicalhan** [comment]
> @logicalhan I think @deads2k point is that putting the global `DefaultFeatureGate` in "k8s.io/component-base/featuregate" instead of "k8s.io/apiserver/pkg/util/feature" makes it a bit more "official".

I see. I'm not sure that distinction is a real one though, at least not in practice, because it is already the default pattern. Your changes also don't preclude the option of components initializing non-default gates and passing them around manually.

**Comment by thockin** [comment]
Agree.  What we really want is something like "k8s.io/internal/features.KubeGates" or something

**Comment by logicalhan** [comment]
Also, one more data point, this pattern is basically how prometheus [defaults a global registry](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L56-L57) and exposes global [convenience methods which refer back to the global registry instances](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L171-L178). It makes it easier to use a default configuration of the API, which can be quite convenient and tends to read nicely. But you can, just like this approach, still instantiate local instances of the otherwise-global object and manually pass it around.

**Comment by deads2k** [comment]
> I'm not sure I follow... `NewFeatureGate()` is still exposed, so you can still create new gates and allow for composability.

But with a single "blessed" location in the library providing featuregates, people won't create their own. Meaning that the client-go example I explained would have client-go using the component-base (well, it won't actually due to dep issues, but the point stands).

**Comment by deads2k** [comment]
> Also, one more data point, this pattern is basically how prometheus [defaults a global registry](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L56-L57) and exposes global [convenience methods which refer back to the global registry instances](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L171-L178).

Great example.  From a practical perspective, all consumers used the copy in the library until they hit the problem I've described above.  At which point they have to embark on a costly migration.  If we avoid the instance in the library, we'll avoid the problem.

**Comment by logicalhan** [comment]
> But with a single "blessed" location in the library providing featuregates, people won't create their own. 

But if that logic was true, then [this](https://github.com/kubernetes/kubernetes/blob/5c2fe63591acab508e1ff5771c3aa70268852318/pkg/kubelet/server/server.go#L322-L326) wouldn't be the case.

**Comment by logicalhan** [comment]
> Great example. From a practical perspective, all consumers used the copy in the library until they hit the problem I've described above. At which point they have to embark on a costly migration. If we avoid the instance in the library, we'll avoid the problem.

I mean, the kubelet already uses and leverages the composability that you're saying this will effectively preclude.

**Comment by deads2k** [comment]
> If we want to encourage a per-org "global default" (not really per-binary, since we have a spiderweb of deps that need to find it), we probably shouldn't call it "Default" and it probably should not live under "apiserver"?

I think a useful middle ground is possible.  The "default to this, but make replaceable by caller" as shown in https://github.com/kubernetes/kubernetes/pull/121122 doesn't require us to change granularity, but does allow us 
1. to compose directly if desired
2. change behavior for binaries a consumer does not compile using an env var
3. allow adapting for conflicting library featuregate names
4. allows overriding of default values when composing

**Comment by deads2k** [comment]
> > Great example. From a practical perspective, all consumers used the copy in the library until they hit the problem I've described above. At which point they have to embark on a costly migration. If we avoid the instance in the library, we'll avoid the problem.
> 
> I mean, the kubelet already uses and leverages the composability that you're saying this will effectively preclude.

Notice 
>Kubernetes also makes extensive use of a global metrics registry to register metrics to be exposed. Aside from general shortcomings of global variables, Kubernetes is seeing actual effects of this, causing a number of components to use `sync.Once` or other mechanisms to ensure to not panic, when registering metrics. Instead a metrics registry should be passed to each component in order to explicitly register metrics instead of through `init` methods or other global, non-obvious executions.

from https://github.com/kubernetes/enhancements/blob/fa647b424a0b11316e8e31274e50332b14434f9c/keps/sig-instrumentation/1206-metrics-overhaul/README.md?plain=1#L50

We switched **after** we got burned.  Our package shape here will significantly influence usage.

**Comment by logicalhan** [comment]
> We switched **after** we got burned. Our package shape here will significantly influence usage.

That's actually not true, we didn't actually [even switch (this uses the global registry)](https://github.com/kubernetes/kubernetes/blob/5c2fe63591acab508e1ff5771c3aa70268852318/pkg/kubelet/server/server.go#L295). We just compose new registries when the need arises (i.e. different kubelet metrics endpoints, as it exposes several).

**Comment by thockin** [comment]
> But with a single "blessed" location in the library providing featuregates, people won't create their own.

We already have a blessed place, though this PR does slightly amplify that.

> From a practical perspective, all consumers used the copy in the library until they hit the problem I've described above. At which point they have to embark on a costly migration.

If the choices are a) use a global; b) pass a new argument to every function, then I think (a) wins almost every time.  There's only so much "you might regret this" that we can push.  If you want to go this way, really, let's delete the globals and force binaries to pass it in Context or as direct arguments.  It will be a LOT of plumbing, though.

**Comment by deads2k** [comment]
I'm not arguing against globals. I'm arguing for a global per binary or repo (say the level of k/k today) as opposed to a global at the level of shared library (as this PR has changed it).

**Comment by logicalhan** [comment]
> Notice
> 
> > Kubernetes also makes extensive use of a global metrics registry to register metrics to be exposed. Aside from general shortcomings of global variables, Kubernetes is seeing actual effects of this, causing a number of components to use `sync.Once` or other mechanisms to ensure to not panic, when registering metrics. Instead a metrics registry should be passed to each component in order to explicitly register metrics instead of through `init` methods or other global, non-obvious executions.
> 
> from https://github.com/kubernetes/enhancements/blob/fa647b424a0b11316e8e31274e50332b14434f9c/keps/sig-instrumentation/1206-metrics-overhaul/README.md?plain=1#L50

I mean, yeah I'm familiar with the stance, but it hasn't actually been super problematic in practice except looking a bit icky (what with the global aspect and whatnot). It hasn't actually prevented us from using and initializing new registries when desired. And the same logic I think applies here, we're already *effectively doing* this with feature gates, and this change doesn't really change that, it just makes the API a little more readable. 

**If** we wanted to migrate to injected and locally instantiated gates, it's a more onerous conversion (as @thockin noted). Similarly, there's a reason why we never migrated all the prometheus registries to use locally instantiated ones; namely, it wasn't really necessary for most components and for components like the kubelet that actually wanted to compose locally, you weren't actually precluded from doing so. 

Sorry, I'm not trying to be difficult, I just think this is a nicer API for an already existing pattern and I don't think it stops you from doing the things you want to do later on in client-go.

**Comment by deads2k** [comment]
> I don't think it stops you from doing the things you want to do later on in client-go.

This change takes us from a global per repo or project to "global for everyone relying on this gating mechanism".  While projects could opt-out, prior art suggests that most will not opt-out until they face another problem.  Let's make a library that encourages good choices.

**Comment by deads2k** [comment]
> And the same logic I think applies here, we're already effectively doing this with feature gates, and this change doesn't really change that, it just makes the API a little more readable.

This PR takes us from an instance used for k/k that is replaceable and very clearly NOT the global to be used generally (since they're in different repos) and exchanges it for an API that strongly encourages usage of a single instance embedded in the library. Put the .Enable method in a repo other than the one defining the type and we've got a completely different API and one that is likely to promote usage we'd be happier with.

**Comment by logicalhan** [comment]
> Put the .Enable method in a repo other than the one defining the type and we've got a completely different API and one that is likely to promote usage we'd be happier with.

I hear what you're saying but I'm personally waffling on the tradeoff between the awkward ergonomics of having dual feature gate imports everywhere (which this PR fixes) and the *potential* problems we avoid by having a delineated API between the type and the global, especially since we control the codebase and since the problematic instance is likely to be client-go, which we know in advance is likely to run into this issue and thus can avoid accordingly.

**Comment by deads2k** [comment]
> especially since we control the codebase and since the problematic instance is likely to be client-go, which we know in advance is likely to run into this issue and thus can avoid accordingly.

We don't.  This became painfully clear as we consumed backports of the http2 reset problem.  Kube provides it, but while we own it here, the issue is downstream consumers.  A tradeoff that makes these problems worse for a slightly shorter Enabled is tough to endorse.

Choosing a new location for a global for featuregates is likely worthwhile.  How about where the keys are defined in pkg/features?

**Comment by logicalhan** [comment]
> We don't. This became painfully clear as we consumed backports of the http2 reset problem. Kube provides it, but while we own it here, the issue is downstream consumers. A tradeoff that makes these problems worse for a slightly shorter Enabled is tough to endorse.

Fair enough. 

> Choosing a new location for a global for featuregates is likely worthwhile. How about where the keys are defined in pkg/features?

There is an init function present there which registers a bunch of default feature gates which would then be dragged in with the moved global (the previous location, being separate, avoids this problem).

**Comment by deads2k** [comment]
> There is an init function present there which registers a bunch of default feature gates which would then be dragged in with the moved global (the previous location, being separate, avoids this problem).

This is usage we actually do own.  We can eliminate the init block and explicitly register when our binary is started.

I've created https://github.com/kubernetes/kubernetes/pull/121864 to propose an alternative that allows composition and avoids the "global in shared library we can conflict on" problem.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/121864#issuecomment-1907157776
